app/uk/gov/hmrc/agentsubscriptionfrontend/auth/AuthActions.scala

```diff
@@ -16,83 +16,92 @@
-    enrolls.find(_.key equals "HMRC-AS-AGENT") match {
-      case Some(enroll) if enroll.isActivated => Future successful true
-      case _ => Future successful false
-    }
-
-  protected def isAgentAffinityGroup()(implicit authContext: AuthContext, hc: HeaderCarrier): Future[Boolean] =
-    authConnector.getUserDetails(authContext).map { userDetailsResponse =>
-      val affinityGroup = (userDetailsResponse.json \ "affinityGroup").as[String]
-      affinityGroup == "Agent"
-    }
-
-  private def redirectToNonAgentNextSteps: Result = {
-    mark("Count-Subscription-NonAgent")
-    Redirect(routes.StartController.showNonAgentNextSteps())
+import uk.gov.hmrc.play.bootstrap.config.AuthRedirects
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class Agent(private val enrolments: Set[Enrolment], private val creds: Credentials) {
+
+  def hasIRPAYEAGENT: Option[Enrolment] = enrolments.find(e => e.key == "IR-PAYE-AGENT" && e.isActivated)
+  def hasIRSAAGENT: Option[Enrolment] = enrolments.find(e => e.key == "IR-SA-AGENT" && e.isActivated)
+
+  def authProviderId = creds.providerId
+  def authProviderType = creds.providerType
+}
+
+object Agent {
+
+  object hasHmrcAsAgentEnrolment {
+    def unapply(agent: Agent): Option[Unit] =
+      if (agent.enrolments.exists(_.key == "HMRC-AS-AGENT")) Some(()) else None
+  }
+
+  object hasNonEmptyEnrolments {
+    def unapply(agent: Agent): Option[Unit] =
+      if (agent.enrolments.nonEmpty) Some(()) else None
   }
 }
+
+trait AuthActions extends AuthorisedFunctions with AuthRedirects with Monitoring {
+
+  def continueUrlActions: ContinueUrlActions
+  def appConfig: AppConfig
+
+  def env = appConfig.environment
+  def config = appConfig.configuration
+
+  def withSubscribingAgent[A](body: Agent => Future[Result])(implicit request: Request[A], hc: HeaderCarrier, ec: ExecutionContext): Future[Result] =
+    authorised(AuthProviders(GovernmentGateway) and AffinityGroup.Agent)
+      .retrieve(allEnrolments and credentials) {
+        case enrolments ~ creds =>
+          if (isEnrolledForHmrcAsAgent(enrolments)) {
+            continueUrlActions.extractContinueUrl.map {
+              case Some(continueUrl) =>
+                mark("Count-Subscription-AlreadySubscribed-HasEnrolment-ContinueUrl")
+                Redirect(continueUrl.url)
+              case None =>
+                mark("Count-Subscription-AlreadySubscribed-HasEnrolment-AgentServicesAccount")
+                Redirect(appConfig.agentServicesAccountUrl)
+            }
+          } else {
+            body(new Agent(enrolments.enrolments, creds))
+          }
+      }.recover {
+        case _: UnsupportedAffinityGroup =>
+          mark("Count-Subscription-NonAgent")
+          Redirect(routes.StartController.showNonAgentNextSteps())
+        case _: NoActiveSession =>
+          toGGLogin(if (appConfig.environment.mode.equals(Mode.Dev)) s"http://${request.host}${request.uri}" else s"${request.uri}")
```

Comment:
Not sure about this one, but I think the `NoActiveSession` case here (and on lines 95, 102 below) might be picked up by the `ErrorHandler.resolveError`?